File: Sliding Window/209. Minimum Size Subarray Sum/Solution.py
from typing import List

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        # N : len(nums)
        # Time Complexity: O(N^2)
        # Space Complexity: O(1)
        # n = len(nums)
        # minLen = float('inf')
        # for i in range(n):
        #     currSum = 0
        #     for j in range(i, n):
        #         currSum += nums[j]
        #         if currSum >= target:
        #             minLen = min(minLen, (j - i + 1))
        #             break
        # return minLen if minLen != float('inf') else 0


        # N : len(nums)
        # Time Complexity: O(2N)
        # Space Complexity: O(1)
        n = len(nums)
        l, r = 0, 0
        currSum = 0
        minLen = float('inf')
        while r < n:
            currSum += nums[r]
            while currSum >= target:
                minLen = min(minLen, (r - l + 1))
                currSum -= nums[l]
                l += 1
            r += 1
        return minLen if minLen != float('inf') else 0

File: Sliding Window/209. Minimum Size Subarray Sum/Solution.java
import java.lang.Integer;
import java.lang.Math;

class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        // N : len(nums)
        // Time Complexity: O(N^2)
        // Space Complexity: O(1)
        // int n = nums.length;
        // int minLen = Integer.MAX_VALUE;
        // for (int i=0; i<n; i++){
        //     int currSum = 0;
        //     for (int j=i; j<n; j++){
        //         currSum += nums[j];
        //         if (currSum >= target){
        //             minLen = Math.min(minLen, (j - i + 1));
        //             break;
        //         }
        //     }
        // }
        // return minLen == Integer.MAX_VALUE ? 0 : minLen;

        
        // N : len(nums)
        // Time Complexity: O(2N)
        // Space Complexity: O(1)
        int n = nums.length;
        int minLen = Integer.MAX_VALUE;
        int currSum = 0;
        int l = 0;
        int r = 0;
        while (r < n){
            currSum += nums[r];
            while (currSum >= target){
                minLen = Math.min(minLen, (r - l + 1));
                currSum -= nums[l];
                l++;
            }
            r++;
        }
        return minLen == Integer.MAX_VALUE ? 0 : minLen;
    }
}